drivers: fpga_ice40: fix CS hack; fix `k_usleep` during spinlock

Finomnis opened the pull request:
Fixes #99948.

Two asserts were triggered in this module:
- Holding a spinlock while `k_usleep` triggers an assert in `k_swap`
- All gpio_set must be on valid pin configs, which the CS hack in this module breaks.

Finomnis committed:
drivers: fpga_ice40: Fix assert in CS hack

The ICE40 driver uses a CS hack that effectively 'steals' the CS pin
from the SPI driver during `load` by setting the CS pin id to 0.

This triggered the newly introduced assert during gpio_set.

This change introduces a hotfix for this where, instead of setting the
CS pin to 0, it sets the entire CS to 'hardware based'.

It is still a hack and should be reworked at some point.

Signed-off-by: Martin Stumpf <finomnis@gmail.com>

JarmouniA reviewed (commented):
I believe the problem here is not in the spinlock usage, which is valid in my opinion since this driver could be used in multi-core setups, rather  in the `k_usleep` introduced in 2f6dff5 without taking into consideration the usage of spinlocks. The mistake went unnoticed because the triggered assert was added just recently.
https://docs.zephyrproject.org/latest/kernel/services/smp/smp.html#spinlocks

Finomnis commented:
@JarmouniA Would you mind elaborating on spinlock vs mutex in this scenario? Why would one be picked over the other, what are the advantages/disadvantages?

The lock will be held during the entire 'load', which can be 50ms. During this entire time, nothing else on this chip can happen if you are in a spinlock, if I understand spinlocks correctly, because it completely prevents the scheduler from unscheduling the current thread.

I don't quite understand the rationale for this, because the underlying SPI could be DMA+Interrupt based. Is it really wise to hold a spinlock during a DMA transfer that might signal its termination with an interrupt?

cfriedt commented:
@Finomnis - I think the `k_usleep()` should be converted to a `k_busy_wait()`. That would definitely be an issue for spinlocks.

The reason that spinlocks were used originally was to avoid any scheduling latencies that a mutex might introduce that could affect the bit-bang interface, which is quite sensitive to timing issues.

We want to avoid the thread being rescheduled while loading the ice40 bitstream.

Finomnis commented:
@cfriedt Better like this?

Finomnis commented on drivers/fpga/fpga_ice40_spi.c:
```diff
@@ -33,21 +33,31 @@ static int fpga_ice40_load(const struct device *dev, uint32_t *image_ptr, uint32
+
+	if (previous_cs.cs_is_gpio) {
+		bus.config.cs = (struct spi_cs_control){
+			.cs_is_gpio = false,
```
If it has any effect in the first place ... I'm unsure. Probably SPI driver dependent. I'll copy it properly to keep the behaviour consistent to how it was before.

Finomnis committed:
drivers: fpga_ice40: rework spinlock usage

The driver uses spinlocks for synchronization (due to strict timing
requirements) which makes it defacto incompatible with almost all SPI
backend drivers, as most of them use `k_sem_take` internally.

This change reworks the driver in the following way to solve this
problem:

- Replace driver locks with `k_mutex`
- Run the `load` functions on the system workqueue to execute them on
  a cooperative priority
- Use additional local spinlock in the `bitbang` driver variant to ensure
  precise timing

Signed-off-by: Martin Stumpf <finomnis@gmail.com>

Finomnis commented:
Still getting this with the current state of the PR:

```
ASSERTION FAIL [arch_irq_unlocked(key) || _kernel.cpus[0].current->base.thread_state & (((1UL << (0))) | ((1UL << (3))))] @ WEST_TOPDIR/zephyr/kernel/include/kswap.h:210
        Context switching while holding lock!
[00:00:00.014,000] <err> os: r0/a1:  0x00000004  r1/a2:  0x000000d2  r2/a3:  0x00000003
[00:00:00.014,000] <err> os: r3/a4:  0x00000004 r12/ip:  0x00000087 r14/lr:  0x3809b583
[00:00:00.014,000] <err> os:  xpsr:  0x01000000
[00:00:00.014,000] <err> os: Faulting instruction address (r15/pc): 0x3802ccb0
[00:00:00.014,000] <err> os: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
[00:00:00.014,000] <err> os: Current thread: 0x20301f00 (main)
[00:00:00.037,000] <err> os: Halting system
```

cfriedt commented:
Kconfig to select mutex or spinlock seems like maybe overkill.

I kind of wonder if there is a way to perform bitstream loading in a high priority or perhaps cooperative workqueue thread.

cfriedt commented:
@Finomnis - you could theoretically use the system work queue, and I believe it's cooperative already.

benediktibk reviewed (approved):
I like the idea, but unfortunately I don't have the STM32 based hardware at the moment available to test it out. I hope to get my hands on it within the next few months, but right now I can only hope it won't break the fickle ICE40 loading procedure.

Finomnis commented:
> I like the idea, but unfortunately I don't have the STM32 based hardware at the moment available to test it out. I hope to get my hands on it within the next few months, but right now I can only hope it won't break the fickle ICE40 loading procedure.

It's pretty broken already 🤷

cfriedt reviewed (commented):
Suggestions to move gratuitous renames to separate commit.

Finomnis commented on drivers/fpga/fpga_ice40_common.c:
```diff
@@ -96,6 +94,15 @@ int fpga_ice40_init(const struct device *dev)
 	const struct fpga_ice40_config *config = dev->config;
+	struct fpga_ice40_data *data = dev->data;
+
+	data->loaded = false;
```
@cfriedt Is it being initialized somewhere? Or should I keep the lines?

mathieuchopstm reviewed (commented):
Agree with @cfriedt - some cosmetic changes are mixed with functional changes, which is not ideal. LGTM otherwise.

mathieuchopstm reviewed (commented):
> > Agree with @cfriedt - some cosmetic changes are mixed with functional changes, which is not ideal. LGTM otherwise.
> 
> @mathieuchopstm Would you mind pointing out which ones? I already reverted the changes @cfriedt was talking about.

These two renames caught my eyes (they generate noise in the diff):
